src/components/pages/variables.jsx

```diff
@@ -1,28 +1,55 @@
 import DocumentTitle from "react-document-title"
 import React, {PropTypes} from "react"
 import {Link} from "react-router"
-import {append, intersperse, lensPath, over, prepend, reduce, sortBy, toPairs} from "ramda"
+import {all, append, intersperse, isNil, lensPath, map, merge, over, pipe, prepend, reduce, reject, sortBy, toPairs, values} from "ramda"
 import TreeView from 'react-treeview'
 
 import * as AppPropTypes from "../../app-prop-types"
 import BreadCrumb from "../breadcrumb"
 import GitHubLink from "../github-link"
 import List from "../list"
+import SearchBox from "../search-box"
 
 
-function buildVariablesTree(variables) {
+function buildVariablesTree(variables, nameFilter, labelFilter) {
   const getVariablePath = variable => prepend('children', intersperse("children", variable.source_file_path.split("/")))
-  return reduce(
-    (memo, variable) => over(
-      lensPath(getVariablePath(variable)),
-      node => node
-        ? over(lensPath(["variables"]), append(variable), node)
-        : {variables: [variable]},
-      memo,
-    ),
-    {},
-    variables
+  const matchFilters = variable => (!nameFilter || variable.name.toLowerCase().includes(nameFilter.toLowerCase())) && (
+    !labelFilter || variable.label && variable.label.toLowerCase().includes(labelFilter.toLowerCase())
   )
+  function removeEmptyNodes(node) {
+    if (isNil(node.children)) {
+      return node
+    } else {
+      const cleanNode = merge(
+        node,
+        {
+          children: pipe(
+            reject(isNil),
+            map(removeEmptyNodes),
+            reject(isNil),
+            )(node.children)
+        }
+      )
+      return all(isNil, values(cleanNode.children))
+        ? null
+        : cleanNode
+    }
+  }
+  return pipe(
+    reduce(
+      (memo, variable) => over(
+        lensPath(getVariablePath(variable)),
+        node => matchFilters(variable)
+          ? node
+            ? over(lensPath(["variables"]), append(variable), node)
+            : {variables: [variable]}
+          : node,
+        memo,
+      ),
+      {}
+    ),
+    removeEmptyNodes,
+  )(variables)
```

Comment:
> Why is it so complicated to display a list in React?

This is not related to React.

This function filters the tree leaves, keeping the intermediary non-empty nodes.
It works on data, not the DOM.

Then this data is passed to the DOM in the [`render` method](https://github.com/openfisca/legislation-explorer/pull/35/files/4977ca339fe53affb2f56f32f3c2262d5c8a2f03#diff-d761e4c762b02dbf437c932af71cb9f0R76).

It is expressed using the Ramda library functions (which is like lodash/underscore) and the `pipe` function chains 2 transformations of data. Perhaps this style is complicated but that's a matter of style, and I can rewrite it in a less functional way.

Anyway I'm gonna post a screenshot of the filtered tree to better understand what I did.